src/backend/distributed/commands/index.c

```diff
@@ -120,103 +127,121 @@ List *
+	 * We first check whether a distributed relation is affected. For that,
+	 * we need to open the relation. To prevent race conditions with later
+	 * lookups, lock the table.
+	 *
+	 * XXX: Consider using RangeVarGetRelidExtended with a permission
+	 * checking callback. Right now we'll acquire the lock before having
+	 * checked permissions, and will only fail when executing the actual
+	 * index statements.
 	 */
-	if (createIndexStatement->relation != NULL)
+	LOCKMODE lockMode = GetCreateIndexRelationLockMode(createIndexStatement);
+	Relation relation = table_openrv(relationRangeVar, lockMode);
+
+	/*
+	 * Before we do any further processing, fix the schema name to make sure
+	 * that a (distributed) table with the same name does not appear on the
+	 * search_path in front of the current schema. We do this even if the
+	 * table is not distributed, since a distributed table may appear on the
+	 * search_path by the time postgres starts processing this command.
+	 */
+	if (relationRangeVar->schemaname == NULL)
 	{
-		LOCKMODE lockmode = ShareLock;
-		MemoryContext relationContext = NULL;
+		/* ensure we copy string into proper context */
+		MemoryContext relationContext = GetMemoryChunkContext(relationRangeVar);
+		char *namespaceName = RelationGetNamespaceName(relation);
+		relationRangeVar->schemaname = MemoryContextStrdup(relationContext,
+														   namespaceName);
+	}
 
-		/*
-		 * We don't support concurrently creating indexes for distributed
-		 * tables, but till this point, we don't know if it is a regular or a
-		 * distributed table.
-		 */
-		if (createIndexStatement->concurrent)
-		{
-			lockmode = ShareUpdateExclusiveLock;
-		}
+	table_close(relation, NoLock);
 
-		/*
-		 * XXX: Consider using RangeVarGetRelidExtended with a permission
-		 * checking callback. Right now we'll acquire the lock before having
-		 * checked permissions, and will only fail when executing the actual
-		 * index statements.
-		 */
-		Relation relation = table_openrv(createIndexStatement->relation, lockmode);
-		Oid relationId = RelationGetRelid(relation);
+	Oid relationId = CreateIndexStmtGetRelationId(createIndexStatement);
+	if (!IsCitusTable(relationId))
+	{
+		return NIL;
+	}
 
-		bool isCitusRelation = IsCitusTable(relationId);
+	ErrorIfUnsupportedIndexStmt(createIndexStatement);
 
-		if (createIndexStatement->relation->schemaname == NULL)
-		{
-			/*
-			 * Before we do any further processing, fix the schema name to make sure
-			 * that a (distributed) table with the same name does not appear on the
-			 * search path in front of the current schema. We do this even if the
-			 * table is not distributed, since a distributed table may appear on the
-			 * search path by the time postgres starts processing this statement.
-			 */
-			char *namespaceName = get_namespace_name(RelationGetNamespace(relation));
+	if (IndexAlreadyExists(createIndexStatement))
+	{
+		/* let standard_processUtility to error out */
+		return NIL;
```

Comment:
I'm guessing we also handle IF NOT EXISTS for an existing index via this code path? Might be worth adding a comment.

Reply:
yeah, I should add a comment for that